### PR TITLE
Add support for uploading files for importing directly from the import screen

### PR DESCRIPTION
**Pull request by cheif**
This PR add support for uploading documents directly into the import folder from the import screen, I've used the standard `<input type="file">` element, so this should be usable on mobile/desktop, and supports multiple files.

Fixes: #904

image

### CODE REVIEW

**Review by yagebu — COMMENTED** [review]
Thanks for the PR, this looks great :)

However, this svelte component is not yet really reload-safe. The `$: files = date` line will overwrite any changes the user made to the file names or accounts. Could you try to see if that can be avoided?

This PR would also get bonus points for allowing drag-and-drop ;)

**Comment by cheif** [comment]
> Thanks for the PR, this looks great :)
> 
> However, this svelte component is not yet really reload-safe. The `$: files = date` line will overwrite any changes the user made to the file names or accounts. Could you try to see if that can be avoided?

Great feedback, I tried to make sure the we don't overwrite `files`, rather just append new entries to it. I'm not sure if it's a reasonable approach though, this is my first time dealing with Svelte. 😅 
> 
> This PR would also get bonus points for allowing drag-and-drop ;)

AFAIK the `<input type="file">` component already supports drag and drop (although I guess that would depend on the browser as well), so I think that's covered already 😄

File: frontend/src/reports/import/Import.svelte [hunk 1]
```diff
@@ -39,9 +45,14 @@
...
 
   onMount(() => router.addInteruptHandler(preventNavigation));
-  onMount(() => {
-    files = data;
-  });
+  $: {
+    // Since the user can change e.g. account/name for imported files, we only add any new files here.
+    // This avoids overwriting the users changes
+    const existingFileNames = files.map((file) => file.name);
+    files = files.concat(
```
Comment (yagebu) [inline frontend/src/reports/import/Import.svelte:52]: I think we want this to be `files = data.map(...);` to ensure that we have a consistent order of files and also correctly handle removal of files. For each file in `data`, we should check whether it already exists in `files` and if so, take that object instead. [Or, to also correctly handle changing importers, not take the whole existing object from `files` but go over all importers and check whether for that given file.name,importer_name combination we already have an `account` and/or `newName` and copy those over]

**Review by yagebu — COMMENTED** [review]
> Great feedback, I tried to make sure the we don't overwrite files, rather just append new entries to it. I'm not sure if it's a reasonable approach though, this is my first time dealing with Svelte.

Yes, definitely the right approach, but see my comment for a minor adjustment that I think would make this more robust. Great first-time Svelte work!

> AFAIK the <input type="file"> component already supports drag and drop (although I guess that would depend on the browser as well), so I think that's covered already

That's true - I was thinking of a more elaborate drag-and-drop where the file would be uploaded directly (and the whole filelist `<div>` could be a drop target). However that might not really be worth the effort (in particular since that means adding a couple of different event handlers for all the drag* events)

File: frontend/src/reports/import/Import.svelte (hunk 1 above)
Comment (cheif) [reply frontend/src/reports/import/Import.svelte]: That's a great suggestion!

I looked at this and took some sort of middle-ground where I just used the `importers` array from the existing entry (if there is one). I think this should cover most cases, I guess in theory the list of importers might have changed, but I guess that won't happen i practice? I think it made the code a bit easier to follow at least.

To get this working I also had to switch to use `router.reload()` when moving and deleting files, I've verified that this works now, and I guess the only reason it wasn't used before was due to this reload-logic not being in place?

**Comment by cheif** [comment]
@yagebu Any more comments on this, or do you think we can get it merged? 😀

File: src/fava/core/ingest.py [hunk 2]
```diff
@@ -184,3 +188,28 @@ def extract(self, filename: str, importer_name: str) -> Entries:
             )
 
         return new_entries_list[0][1]
+
+
+def filepath_in_primary_imports_folder(
```
Comment (yagebu) [inline src/fava/core/ingest.py:193]: This function needs to replace any separators in the given filname like `filepath_in_document_folder` does. Will fix that in the merge.

**Review by yagebu — APPROVED** [review]

**Comment by sumanchapai** [comment]
Why was this PR closed? I have been looking for a feature like this in beancount where I can upload invoices but not sure if this feature is available. I would also be open to contributing if it doesn't exist yet.

**Pull request closed without merging**